qiskit/pulse/builder.py

```diff
@@ -740,6 +744,16 @@ def append_block(self, context_block: ScheduleBlock):
         if len(context_block) > 0:
             self._context_stack[-1].append(context_block)
 
+    def append_reference(self, *ref_keys: str, name: Optional[str] = None):
+        """Add external program as a :class:`~qiskit.pulse.instructions.Reference` instruction.
+
+        Args:
+            ref_keys: Set of keys to uniquely specify the reference.
+            name: Optional. A label of instruction.
+        """
+        reference = instructions.Reference(*ref_keys, name=name)
+        self.append_instruction(reference)
```

Comment:
Here I have a question: Should this be doing any validation? This could be needed for nested references, e.g.,`root::y1(d0)::x1(d0)` seems to make sense (referenced `y1` on channel `d0` references `x1` played on channel `d0`, which is consistent) but `root::y1(d0)::x1(d1)` does not (referenced `y1` on channel `d0` references `x1` played on channel `d1`, which is not consistent).

Reply:
As @wshanks suggested, reference is not aware of channels. This allows us to avoid complicated assignment problems, i.e. channel or qubit might be parameterized. So channels of the reference is treated unknown until assignment. But you can provide some information about channel or qubit in key tuples on your responsibility, i.e. that management is delegated to calibrations.